### ISSUE TITLE
can't use the provider with my argocd (PROTOCOL_ERROR)

### ISSUE THREAD

**Issue by zerr0s** [oboukili/terraform-provider-argocd#217]
Hello,

I'm trying to use this provider with argocd 2.5.2.

```
 Error: Failed to init clients
│ 
│   with argocd_repository_credentials.xxxx-gitlab-credentials,
│   on argocd.tf line 14, in resource "argocd_repository_credentials" "xxxx-gitlab-credentials":
│   14: resource "argocd_repository_credentials" "xxxx-gitlab-credentials" {
│ 
│ rpc error: code = Internal desc = stream terminated by RST_STREAM with error code: PROTOCOL_ERROR
```

My config is simple:
```
provider "argocd" {
  server_addr = base64decode(var.argocd_endpoint)
  auth_token  = base64decode(var.argocd_token)
  grpc_web    = true
}
```

The cli is working well:
`argocd --grpc-web app list`

I don't understand what I am missing. Any idea please ?

**Comment by zerr0s** [comment]
It seems that http2 is supported on my traefik, the LB of my cloud provider.

So I have test  directly the argocd service directly inside my clusterwith curl with the --htt2 option to force to use http2.
argocd is responding with http/1.1

I'm using the helm chart and I am not finding any option to force to use http2. So i don't know how to resolve this issue with this terraform module.

**Comment by mrcswz** [comment]
Hi @zerr0s, have you fixed this meanwhile? 

Unfortunately I'm facing a similar error and have no idea on how to fix this.

**Comment by viorelvintan** [comment]
Hi @zerr0s and @mrcswz ,

I am facing the same error with provider version 5.2.0 when attempting to create a resource (in my case an `argocd_application`).
As to the server version, I have tried with both 2.6.7 and 2.7.1, same result, even though `argocd --grpc-web app list` does work.

```
argocd_application.app_of_apps: Creating...
│ Error: failed to init clients
│
│   with argocd_application.app_of_apps,
│   on main.tf line 44, in resource "argocd_application" "app_of_apps":
│   44: resource "argocd_application" "app_of_apps" {
│
│ rpc error: code = Internal desc = stream terminated by RST_STREAM with error code: PROTOCOL_ERROR
❯ argocd --grpc-web app list
NAME  CLUSTER  NAMESPACE  PROJECT  STATUS  HEALTH  SYNCPOLICY  CONDITIONS  REPO  PATH  TARGET
```

Have you been able to get past this issue?

**Comment by onematchfox** [comment]
Hi folks,

Please check out the [reports of the same issue](https://github.com/search?q=repo%3Aargoproj%2Fargo-cd+stream+terminated+by+RST_STREAM+with+error+code%3A+PROTOCOL_ERROR&type=issues) within the `argoproj/argo-cd`. From my point of view, it looks like this error is indicative of a configuration issue (e.g. an issue with the `auth_token` configured on the provider). It's also possible that the original author's issue was resolved by #10704?

As we are unable to reproduce the issue locally, unfortunately, there's not much we can do to help. If anyone can provide a minimum reproducible example, then I'd be more than happy to take a look. 

Also, it might be worth trying out one of the other mechanisms to connect to ArgoCD (e.g. using a `username/password` instead of supplying an `auth_token` or setting `use_local_config = true` if you have already run `argocd login` and CLI commands are working. This would at least narrow down the options for what might be causing the problem.

**Comment by oboukili** [comment]
It’s a network issue with regards to how you configured your frontend loadbalancer, proxying argocd-server. By default , most HTTP (L7) reverse proxies terminate the http2 client connection, and proxy the request using http1.1 to the backend, most of them can also proxy the request using http2.
the argocd client library (used in both the cli and this terraform provider) expects the requests to be passed through (which is easily achievable using a L4 network proxy instead).

The grpc-web client option (also available with this provider) circumvents this issue, unfortunately, from experience, this is quite flaky, especially for bidirectional sync ops.

this is out of the scope of support for this terraform provider, and this should be tackled upstream, at the argocd level.
